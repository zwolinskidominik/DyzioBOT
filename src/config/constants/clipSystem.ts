export const JURY_ROLE_ID = '1303735601845239969';
export const OWNER_ROLE_ID = '881295973782007868';
export const CLIPS_CHANNEL_ID = '1210246300843647047';

export const SEPARATOR_GIF_URL =
  'https://64.media.tumblr.com/64084f352d1664758e1a4febcb0e4464/8ac72bb49761ea20-51/s500x750/0646a52b0686cac841fd1201edde418ccddf6443.gif';

export const REACTION_TO_SCORE: Record<string, number> = {
  '1️⃣': 1,
  '2️⃣': 2,
  '3️⃣': 3,
  '4️⃣': 4,
  '5️⃣': 5,
  '6️⃣': 6,
};

export const VALID_REACTIONS = Object.keys(REACTION_TO_SCORE);
