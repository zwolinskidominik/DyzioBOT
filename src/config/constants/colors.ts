export const COLORS = {
  CS2_MIX: '#7C000C',
  DEFAULT: '#4C4C54',
  BIRTHDAY: 'EA596E',
  EMBED: '#4C4C54',
  ERROR: '#E74D3C',
  FACEIT: '#FF5500',
  FORTUNE: '#AA8DD8',
  FORTUNE_ADD: '#00FF00',
  GIVEAWAY: '#5865F2',
  GIVEAWAY_ENDED: '#4C4C54',
  JOIN: '#86C232',
  LEAVE: '#FF0000',
  MEME: '#4C4C54',
  TICKET: '#5865F2',
  TICKET_REPORT: '#ED4245',
  TICKET_PARTNERSHIP: '#FEE75C',
  TICKET_IDEA: '#57F287',
  TICKET_REWARD: '#7C1196',
  TWITCH: '#9146FF',
  WARN: '#F1C40F',
  WARNINGS_LIST: '#FFD700',
} as const;
