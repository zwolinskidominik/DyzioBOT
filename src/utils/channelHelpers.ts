import { Collection, Guild, GuildMember, TextChannel } from 'discord.js';
import { ChannelStatsModel } from '../models/ChannelStats';
import logger from './logger';

// Track in-flight rename intentions per channel to ensure last-write-wins and avoid double retries
const renameTokens = new Map<string, number>();
let renameSeq = 0;

export async function safeSetChannelName(
  channel: TextChannel,
  newName: string,
  retries = 3,
  delay = 1_000
): Promise<void> {
  if (channel.name === newName) return;
  const chanId = (channel as any).id ?? '';
  const myToken = ++renameSeq;
  renameTokens.set(chanId, myToken);

  const attempt = async (left: number, nextDelay: number): Promise<void> => {
    // Abort if a newer rename request exists for this channel
    if (renameTokens.get(chanId) !== myToken) return;
    try {
      await channel.setName(newName);
    } catch (error: unknown) {
      const e = error as { httpStatus?: number; code?: number };
      if (e.code === 50_013) {
        logger.warn(`Brak uprawnień do zmiany nazwy kanału ${channel.id}`);
        return;
      }
      if (left > 0 && e?.httpStatus === 429) {
        logger.warn(
          `Rate-limit przy zmianie nazwy kanału ${channel.id}. Retry za ${nextDelay} ms (pozostało ${left}).`
        );
        await new Promise((r) => setTimeout(r, nextDelay));
        // check again before retrying to avoid double-retry overriding a newer request
        if (renameTokens.get(chanId) !== myToken) return;
        return attempt(left - 1, Math.min(nextDelay * 2, 30_000));
      }
      throw error;
    }
  };

  await attempt(retries, delay);
}

export interface SimpleChannelConfig {
  channelId?: string;
  template?: string;
  member?: string;
}

function buildChannelName(template: string | undefined, value: string | number): string {
  if (!template) return String(value);
  return template.replace(/<>|{value}/g, String(value)).slice(0, 100);
}

export async function updateChannelName(
  guild: Guild,
  channelConfig: SimpleChannelConfig | undefined,
  newValue: string | number
): Promise<void> {
  if (!channelConfig?.channelId) return;
  const channel = guild.channels.cache.get(channelConfig.channelId);
  if (
    !channel ||
    typeof (channel as any).setName !== 'function' ||
    typeof (channel as any).name !== 'string'
  )
    return;

  const newName = buildChannelName(channelConfig.template, newValue);
  await safeSetChannelName(channel as TextChannel, newName);
}

const BAN_CACHE_TTL = 10 * 60 * 1000;
const banCache = new Map<string, { count: number; timestamp: number }>();

const MEMBER_FETCH_TTL = 5 * 60 * 1000;
const lastMemberFetch = new Map<string, number>();

async function ensureFreshMembers(guild: Guild): Promise<void> {
  const incomplete = guild.members.cache.size < guild.memberCount;
  const now = Date.now();
  const last = lastMemberFetch.get(guild.id) ?? 0;

  if (!incomplete) return;
  if (now - last < MEMBER_FETCH_TTL) return;

  try {
    await guild.members.fetch();
    lastMemberFetch.set(guild.id, now);
  } catch (err) {
    logger.warn(`Nie udało się pobrać pełnej listy członków guild=${guild.id}: ${err}`);
  }
}

function computeNewestNonBotMember(
  members: Collection<string, GuildMember>
): GuildMember | undefined {
  let newest: GuildMember | undefined;
  let newestTs = -1;
  for (const m of members.values()) {
    if (m.user.bot) continue;
    const ts = m.joinedTimestamp ?? 0;
    if (ts > newestTs) {
      newestTs = ts;
      newest = m;
    }
  }
  return newest;
}

async function getBanCount(guild: Guild): Promise<number> {
  const cached = banCache.get(guild.id);
  const now = Date.now();
  if (cached && now - cached.timestamp < BAN_CACHE_TTL) return cached.count;
  try {
    const bans = await guild.bans.fetch();
    const count = bans.size;
    banCache.set(guild.id, { count, timestamp: now });
    return count;
  } catch (err) {
    logger.error(`Błąd przy pobieraniu banów dla guild=${guild.id}: ${err}`);
    return cached?.count ?? 0;
  }
}

export async function updateChannelStats(guild: Guild): Promise<void> {
  const channelStats = await ChannelStatsModel.findOne({ guildId: guild.id });
  if (!channelStats) return;

  try {
    await ensureFreshMembers(guild);

    const membersCache = guild.members.cache;
    const nonBotMembers = membersCache.filter((m) => !m.user.bot);
    const botMembers = membersCache.filter((m) => m.user.bot);

    const userCount = nonBotMembers.size;
    const botCount = botMembers.size;

    const newestMember = computeNewestNonBotMember(nonBotMembers);
    const newestValue = newestMember?.user.username ?? 'Brak';

    const banCount = await getBanCount(guild);

    const tasks: Promise<void>[] = [];

    if (channelStats.channels.users) {
      tasks.push(updateChannelName(guild, channelStats.channels.users, userCount));
    }
    if (channelStats.channels.bots) {
      tasks.push(updateChannelName(guild, channelStats.channels.bots, botCount));
    }
    if (channelStats.channels.bans) {
      tasks.push(updateChannelName(guild, channelStats.channels.bans, banCount));
    }
    if (channelStats.channels.lastJoined) {
      tasks.push(
        updateChannelName(guild, channelStats.channels.lastJoined, newestValue).then(() => {
          channelStats.channels.lastJoined!.member = newestMember?.id;
        })
      );
    }

    await Promise.all(tasks);
    await channelStats.save();
  } catch (err) {
    logger.error(`Błąd przy aktualizacji statystyk (guild=${guild.id}): ${err}`);
  }
}
