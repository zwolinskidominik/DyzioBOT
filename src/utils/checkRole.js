/**
 * Funkcja pomocnicza checkRole sprawdza, czy targetMember (użytkownik docelowy)
 * może być moderowany przez requestMember (użytkownika wywołującego komendę)
 * przy jednoczesnej weryfikacji hierarchii bota (botMember).
 *
 * Warunki:
 * - Nie można moderować właściciela serwera.
 * - Rola targetMember musi mieć niższą pozycję niż rola requestMember.
 * - Rola targetMember musi mieć niższą pozycję niż rola bota.
 *
 * @param {GuildMember} targetMember - Użytkownik, którego chcemy moderować.
 * @param {GuildMember} requestMember - Użytkownik, który wywołuje komendę.
 * @param {GuildMember} botMember - Członek bota (zazwyczaj interaction.guild.members.me).
 * @returns {boolean} - Zwraca true, gdy targetMember może być moderowany, w przeciwnym razie false.
 */
function checkRole(targetMember, requestMember, botMember) {
    if (!targetMember || !requestMember || !botMember) {
      throw new Error("Wszystkie trzy parametry muszą być dostarczone.");
    }
    // Nie można moderować właściciela serwera
    if (targetMember.id === targetMember.guild.ownerId) {
      return false;
    }
    const targetPos = targetMember.roles.highest.position;
    const requestPos = requestMember.roles.highest.position;
    const botPos = botMember.roles.highest.position;
  
    if (targetPos >= requestPos) return false;
    if (targetPos >= botPos) return false;
    return true;
  }
  
  module.exports = checkRole;
  