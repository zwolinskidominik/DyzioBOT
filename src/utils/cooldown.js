const globalCooldowns = new Map();
const COOLDOWN_TIME = 2500;

/**
 * Sprawdza, czy użytkownik jest na cooldownie.
 * Jeśli tak, zwraca liczbę sekund, które pozostały do końca cooldownu.
 * Jeśli nie, ustawia cooldown dla użytkownika i zwraca 0.
 *
 * @param {string} userId - ID użytkownika
 * @returns {number} - Liczba sekund pozostałych do końca cooldownu (0, jeśli cooldown wygasł)
 */
function checkGlobalCooldown(userId) {
  const now = Date.now();
  if (globalCooldowns.has(userId)) {
    const expirationTime = globalCooldowns.get(userId);
    if (now < expirationTime) {
      return (expirationTime - now) / 1000;
    }
  }
  globalCooldowns.set(userId, now + COOLDOWN_TIME);
  return 0;
}

module.exports = { checkGlobalCooldown };
