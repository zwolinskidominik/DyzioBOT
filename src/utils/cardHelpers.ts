import { Builder, loadImage, JSX } from 'canvacord';
import type { ICardOptions, IGreetingType } from '../interfaces/Cards';
import { COLORS } from '../config/constants/colors';
import * as path from 'path';

const BLANK_PNG = path.resolve(__dirname, '../../assets/bg.jpg');

export class GreetingsCard extends Builder {
  private opts: ICardOptions;

  constructor() {
    super(930, 280);

    this.opts = {
      type: 'welcome',
      displayName: '',
      avatar: BLANK_PNG,
      message: '',
      backgroundImage: path.resolve(__dirname, '../../assets/bg.jpg'),
    };

    this.bootstrap(this.opts as unknown as Record<string, unknown>);
  }

  public setType(v: IGreetingType): this {
    this.opts.type = v;
    this.options.set('type', v);
    return this;
  }

  public setDisplayName(v: string): this {
    this.opts.displayName = v;
    this.options.set('displayName', v);
    return this;
  }

  public setAvatar(v: string | null | undefined): this {
    this.opts.avatar = v || BLANK_PNG;
    this.options.set('avatar', this.opts.avatar);
    return this;
  }

  public setMessage(v: string): this {
    this.opts.message = v;
    this.options.set('message', v);
    return this;
  }

  private static async safeLoad(src: string) {
    try {
      if (src.includes('cdn.discordapp.com')) {
        return await this.loadWithRetry(src);
      }

      if (src.startsWith('/') || src.includes(':\\')) {
        return await loadImage(BLANK_PNG);
      }

      return await loadImage(src);
    } catch (error) {
      console.error(`Failed to load image from ${src}:`, error);
      return await loadImage(BLANK_PNG);
    }
  }

  private static async loadWithRetry(src: string, retries = 3) {
    let lastError;

    for (let attempt = 0; attempt < retries; attempt++) {
      try {
        return await loadImage(src);
      } catch (error) {
        lastError = error;
        await new Promise((resolve) => setTimeout(resolve, 1000 * (attempt + 1)));
      }
    }

    console.error(`All retries failed for ${src}:`, lastError);
    return await loadImage(BLANK_PNG);
  }

  public async render(): Promise<JSX.Element> {
    const { type, displayName, avatar, message, backgroundImage } = this.opts;

    let avatarImg, bgImg;
    try {
      [bgImg, avatarImg] = await Promise.all([
        GreetingsCard.safeLoad(backgroundImage),
        GreetingsCard.safeLoad(avatar),
      ]);
    } catch (error) {
      console.error('Error loading images:', error);
      try {
        bgImg = await GreetingsCard.safeLoad(backgroundImage);
      } catch {
        bgImg = await loadImage(BLANK_PNG);
      }

      try {
        avatarImg = await GreetingsCard.safeLoad(avatar);
      } catch {
        avatarImg = await loadImage(BLANK_PNG);
      }
    }

    const accent = type === 'welcome' ? COLORS.JOIN : COLORS.LEAVE;
    const greet = type === 'welcome' ? 'Cześć' : 'Żegnaj';

    return JSX.createElement(
      'div',
      {
        className:
          'h-full w-full flex flex-col items-center justify-center bg-[#23272A] rounded-xl',
        style: {
          backgroundImage: `url(${bgImg.toDataURL()})`,
          backgroundSize: 'cover',
          backgroundRepeat: 'no-repeat',
        },
      },
      JSX.createElement(
        'div',
        {
          className: 'px-6 bg-[#2B2F35] w-[96%] h-[84%] rounded-lg flex items-center',
          style: { opacity: 0.95 },
        },
        JSX.createElement('img', {
          src: avatarImg.toDataURL(),
          className: 'h-[40] w-[40] rounded-full',
        }),
        JSX.createElement(
          'div',
          { className: 'flex flex-col ml-6' },
          JSX.createElement(
            'h1',
            { className: 'text-5xl text-white font-bold m-0' },
            `${greet}, `,
            JSX.createElement('span', { style: { color: accent } }, `${displayName}!`)
          ),
          JSX.createElement('p', { className: 'text-gray-300 text-3xl m-0 mt-2' }, message)
        )
      )
    );
  }
}
