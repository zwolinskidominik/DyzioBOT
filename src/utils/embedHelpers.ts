import { EmbedBuilder, ColorResolvable } from 'discord.js';
import { IBaseEmbedOptions } from '../interfaces/Embed';
import { getGuildConfig } from '../config/guild';
import { COLORS } from '../config/constants/colors';

export function createBaseEmbed(options: IBaseEmbedOptions = {}): EmbedBuilder {
  const {
    isError = false,
    color = '',
    title = '',
    description = '',
    footerText = '',
    footerIcon = '',
    image = '',
    thumbnail = '',
    authorName = '',
    authorIcon = '',
    authorUrl = '',
    url = '',
    timestamp = false,
  } = options;

  const finalColor = color || (isError ? COLORS.ERROR : COLORS.DEFAULT);
  const embed = new EmbedBuilder().setColor(finalColor as ColorResolvable);
  if (timestamp) embed.setTimestamp();

  if (title) embed.setTitle(title);
  if (description) embed.setDescription(description);
  if (footerText) {
    embed.setFooter({ text: footerText, iconURL: footerIcon || undefined });
  }
  if (image) embed.setImage(image);
  if (thumbnail) embed.setThumbnail(thumbnail);
  if (authorName) {
    embed.setAuthor({
      name: authorName,
      iconURL: authorIcon || undefined,
      url: authorUrl || undefined,
    });
  }
  if (url) embed.setURL(url);

  return embed;
}

export function formatResults(
  guildId: string,
  upvotes: string[] = [],
  downvotes: string[] = []
): string {
  const {
    emojis: { suggestionPB },
  } = getGuildConfig(guildId);

  const total = upvotes.length + downvotes.length;
  const length = 14;
  const filled = total ? Math.round((upvotes.length / total) * length) : 0;
  const empty = length - filled;
  const upP = total ? (upvotes.length / total) * 100 : 0;
  const downP = total ? (downvotes.length / total) * 100 : 0;

  const bar =
    (filled ? suggestionPB.lf : suggestionPB.le) +
    suggestionPB.mf.repeat(filled) +
    suggestionPB.me.repeat(empty) +
    (filled === length ? suggestionPB.rf : suggestionPB.re);

  return `👍 ${upvotes.length} głosów na tak (${upP.toFixed(1)}%) • 👎 ${downvotes.length} głosów na nie (${downP.toFixed(1)}%)\n${bar}`;
}

export function formatWarnBar(guildId: string, count: number): string {
  const {
    emojis: { warnPB },
  } = getGuildConfig(guildId);

  const length = 9;
  const maxWarnings = 3;
  const filled = Math.round((count / maxWarnings) * length);
  const empty = length - filled;

  const start = filled > 0 ? warnPB.lf : warnPB.le;
  const middleFull = filled > 1 ? warnPB.mf.repeat(filled - 1) : '';
  const middleEmpty = empty > 1 ? warnPB.me.repeat(empty - 1) : '';
  const end = filled === length ? warnPB.rf : warnPB.re;

  return `${start}${middleFull}${middleEmpty}${end}`;
}
