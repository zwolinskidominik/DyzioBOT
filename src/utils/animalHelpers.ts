import { EmbedBuilder, CommandInteraction, MessageFlags } from 'discord.js';
import { IAnimalImageResponse, IAnimalCommandConfig } from '../interfaces/api/Animal';
import { createBaseEmbed } from '../utils/embedHelpers';
import { COLORS } from '../config/constants/colors';
import logger from '../utils/logger';
import { request } from 'undici';

const lastImageId = new Map<string, string>();

export async function fetchRandomAnimalImage(
  config: IAnimalCommandConfig
): Promise<IAnimalImageResponse | null> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5_000);
  try {
    const first = await request(config.apiURL, { signal: controller.signal });
    const json = await first.body.json();
    let arr: IAnimalImageResponse[] = Array.isArray(json) ? json : [json];
    if (!arr.length) {
      logger.warn(`API ${config.apiSource} zwróciło pustą odpowiedź.`);
      return null;
    }
    const prevId = lastImageId.get(config.apiURL);
    let chosen = arr.length > 1 ? arr.find((x) => x.id !== prevId) || arr[0] : arr[0];
    if (prevId && chosen.id === prevId) {
      try {
        const second = await request(config.apiURL, { signal: controller.signal });
        const json2 = await second.body.json();
        const arr2: IAnimalImageResponse[] = Array.isArray(json2) ? json2 : [json2];
        if (arr2.length) {
          const alt = arr2.find((x) => x.id !== prevId) || arr2[0];
          if (alt.id !== prevId) chosen = alt;
        }
      } catch {}
    }
    if (!chosen?.url) return null;
    lastImageId.set(config.apiURL, chosen.id);
    return chosen;
  } catch (error) {
    const aborted = (error as any)?.name === 'AbortError';
    logger.error(
      `Błąd pobierania ${config.animalType} (${config.apiSource})${aborted ? ' (timeout)' : ''}: ${error}`
    );
    return null;
  } finally {
    clearTimeout(timeout);
  }
}

export function createAnimalEmbed(
  animal: IAnimalImageResponse,
  config: IAnimalCommandConfig
): EmbedBuilder {
  return createBaseEmbed({
    title: `Losowy ${config.animalTitle}`,
    footerText: `Utworzone poprzez API z ${config.apiSource} - ID: ${animal.id}`,
    image: animal.url,
    color: COLORS.DEFAULT,
  });
}

export async function handleAnimalError(
  interaction: CommandInteraction,
  config: IAnimalCommandConfig
): Promise<void> {
  await interaction.followUp({
    content: config.errorMessage,
    flags: MessageFlags.Ephemeral,
  });
}
