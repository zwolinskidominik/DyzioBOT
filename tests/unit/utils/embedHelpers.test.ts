// Removed duplicate minimal embed tests to avoid redeclaration; richer tests retained below.
import { createBaseEmbed, formatResults, formatWarnBar } from '../../../src/utils/embedHelpers';
import { COLORS } from '../../../src/config/constants/colors';

const BOT_ID = '1119327417237000285';

describe('embedHelpers.createBaseEmbed', () => {
  test('sets provided fields and omits undefined', () => {
    const embed = createBaseEmbed({
      title: 'T',
      description: 'D',
      footerText: 'F',
      footerIcon: 'http://x/icon.png',
      image: 'http://x/i.png',
      thumbnail: 'http://x/t.png',
      authorName: 'A',
      authorIcon: 'http://x/a.png',
      authorUrl: 'http://x/a',
      url: 'http://x/u',
      timestamp: true,
      color: '#123456',
    }).toJSON();

    expect(embed.title).toBe('T');
    expect(embed.description).toBe('D');
    expect(embed.footer?.text).toBe('F');
    expect(embed.footer?.icon_url).toBe('http://x/icon.png');
    expect(embed.image?.url).toBe('http://x/i.png');
    expect(embed.thumbnail?.url).toBe('http://x/t.png');
    expect(embed.author?.name).toBe('A');
    expect(embed.author?.icon_url).toBe('http://x/a.png');
    expect(embed.author?.url).toBe('http://x/a');
    expect(embed.url).toBe('http://x/u');
    expect(embed.timestamp).toBeDefined();
    expect(embed.color).toBe(parseInt('#123456'.replace('#', ''), 16));
  });

  test('defaults to error color when isError true and no explicit color', () => {
    const embed = createBaseEmbed({ isError: true, title: 'Err' }).toJSON();
    expect(embed.color).toBe(parseInt(COLORS.ERROR.replace('#', ''), 16));
  });

  test('defaults to base color when nothing provided', () => {
    const embed = createBaseEmbed().toJSON();
    expect(embed.color).toBe(parseInt(COLORS.DEFAULT.replace('#', ''), 16));
  });

  test('omits optional fields when falsy and uses DEFAULT when isError false without color', () => {
    const embed = createBaseEmbed({
      title: 'OnlyTitle',
      description: '', // falsy -> omitted
      footerText: '', // falsy -> omitted
      image: undefined,
      thumbnail: undefined,
      authorName: undefined as any,
      url: undefined,
      // isError not set -> DEFAULT color
    }).toJSON();
    expect(embed.title).toBe('OnlyTitle');
    expect(embed.description).toBeUndefined();
    expect(embed.footer).toBeUndefined();
    expect(embed.image).toBeUndefined();
    expect(embed.thumbnail).toBeUndefined();
    expect(embed.author).toBeUndefined();
    expect(embed.url).toBeUndefined();
    expect(embed.color).toBe(parseInt(COLORS.DEFAULT.replace('#', ''), 16));
  });
});

describe('embedHelpers.formatResults', () => {
  test('renders bar and percentages', () => {
    const out = formatResults(BOT_ID, ['u1', 'u2', 'u3', 'u4'], ['d1']);
    expect(out).toMatch(/👍 4 głosów na tak/);
    expect(out).toMatch(/👎 1 głosów na nie/);
    expect(out).toMatch(/%/);
  });

  test('0 votes shows 0% and builds empty bar', () => {
    const out = formatResults(BOT_ID, [], []);
    expect(out).toMatch(/0 głosów na tak/);
    expect(out).toMatch(/0 głosów na nie/);
  });

  test('100% upvotes -> full bar', () => {
    const cfg = (require('../../../src/config/bot') as any).getBotConfig(BOT_ID);
    const { suggestionPB } = cfg.emojis;
    const out = formatResults(BOT_ID, new Array(5).fill(0).map((_, i) => 'u'+i), []);
    expect(out).toMatch(/100\.0%/);
    expect(out).toMatch(/0\.0%/);
    expect(out).toContain(suggestionPB.lf);
    expect(out).toContain(suggestionPB.rf);
    expect(out).not.toContain(suggestionPB.re);
    const mfCount = (out.match(new RegExp(suggestionPB.mf.replace(/[.*+?^${}()|[\]\\]/g,'\\$&'),'g'))||[]).length;
    expect(mfCount).toBeGreaterThanOrEqual(12);
  });

  test('100% downvotes -> empty bar', () => {
    const cfg = (require('../../../src/config/bot') as any).getBotConfig(BOT_ID);
    const { suggestionPB } = cfg.emojis;
    const out = formatResults(BOT_ID, [], ['d1','d2','d3']);
    expect(out).toMatch(/0\.0%/);
    expect(out).toMatch(/100\.0%/);
    expect(out).toContain(suggestionPB.le);
    expect(out).toContain(suggestionPB.re);
    expect(out).not.toContain(suggestionPB.rf);
    const mfCount = (out.match(new RegExp(suggestionPB.mf.replace(/[.*+?^${}()|[\]\\]/g,'\\$&'),'g'))||[]).length;
    expect(mfCount).toBe(0);
  });

  test('empty lists snapshot', () => {
    const out = formatResults(BOT_ID, [], []);
    expect(out).toMatchSnapshot('formatResults-empty');
  });

  test('very long lists snapshot (bar length fixed)', () => {
    const ups = Array.from({ length: 900 }, (_, i) => `u${i}`);
    const downs = Array.from({ length: 100 }, (_, i) => `d${i}`);
    const out = formatResults(BOT_ID, ups, downs);
    expect(out).toMatchSnapshot('formatResults-long');
  });
});

describe('embedHelpers.formatWarnBar', () => {
  test('warn bar increases representation of full segments with count', () => {
    const low = formatWarnBar(BOT_ID, 0);
    const mid = formatWarnBar(BOT_ID, 1);
    const high = formatWarnBar(BOT_ID, 3);
    const { warnPB } = (require('../../../src/config/bot') as any).getBotConfig(BOT_ID).emojis;
    const fullToken = warnPB.mf;
    const countFull = (s: string) =>
      (s.match(new RegExp(fullToken.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g')) || []).length;
    const lowFull = countFull(low);
    const midFull = countFull(mid);
    const highFull = countFull(high);
    expect(highFull).toBeGreaterThanOrEqual(midFull);
    expect(midFull).toBeGreaterThanOrEqual(lowFull);
    expect(new Set([low, mid, high]).size).toBeGreaterThan(1);
  });

  test('warn bar caps at max warnings', () => {
    const capped = formatWarnBar(BOT_ID, 999);
    const max = formatWarnBar(BOT_ID, 3);
    expect(capped).toBe(max);
  });

  test('createBaseEmbed + warn bar snapshot', () => {
    const warnBar = formatWarnBar(BOT_ID, 2);
    const embed = createBaseEmbed({
      isError: true,
      title: 'Status ostrzeżeń',
      description: `Ostrzeżenia: ${warnBar}`,
      footerText: 'Serwer XYZ',
      thumbnail: 'http://assets/thumb.png',
      authorName: 'Dyzio',
      authorIcon: 'http://assets/author.png',
    }).toJSON();
    expect(embed).toMatchSnapshot('createBaseEmbed-with-warnbar');
  });
});
