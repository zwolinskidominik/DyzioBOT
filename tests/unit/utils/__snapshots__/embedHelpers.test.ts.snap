// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedHelpers.formatResults empty lists snapshot: formatResults-empty 1`] = `
"👍 0 głosów na tak (0.0%) • 👎 0 głosów na nie (0.0%)
<:5499lb2g:1299663909040558160><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2827l2g:1299663896218570805><:2881lb3g:1299663884562468874>"
`;

exports[`embedHelpers.formatResults very long lists snapshot (bar length fixed): formatResults-long 1`] = `
"👍 900 głosów na tak (90.0%) • 👎 100 głosów na nie (10.0%)
<:5988lbg:1299663872071831622><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:3451lg:1299663858914295818><:2827l2g:1299663896218570805><:2881lb3g:1299663884562468874>"
`;

exports[`embedHelpers.formatWarnBar createBaseEmbed + warn bar snapshot: createBaseEmbed-with-warnbar 1`] = `
{
  "author": {
    "icon_url": "http://assets/author.png",
    "name": "Dyzio",
    "url": undefined,
  },
  "color": 15158588,
  "description": "Ostrzeżenia: <:yleftfull:1366150886497321143><:ymidfull:1366150872727294042><:ymidfull:1366150872727294042><:ymidfull:1366150872727294042><:ymidfull:1366150872727294042><:ymidfull:1366150872727294042><:ymidempty:1366151054672007249><:ymidempty:1366151054672007249><:yrightempty:1366151038679257350>",
  "footer": {
    "icon_url": undefined,
    "text": "Serwer XYZ",
  },
  "thumbnail": {
    "url": "http://assets/thumb.png",
  },
  "title": "Status ostrzeżeń",
}
`;
